src/crimson/osd/replicated_backend.cc

```diff
@@ -53,7 +53,7 @@ ReplicatedBackend::_submit_transaction(std::set<pg_shard_t>&& pg_shards,
     throw crimson::common::actingset_changed(peering->is_primary);
   }
 
-  const ceph_tid_t tid = next_txn_id++;
+  const ceph_tid_t tid = shard_services.get_tid();
```

Comment:
Is this because the tid needs to be unique across pgs?

Reply:
yes.

Reply:
This isn't actually correct.  The tid on a repop should match the tid from the original client request.  PrimaryLogPG::do_op constructs the OpContext for the op using MOSDOp::get_reqid():

```

  dout(25) << __func__ << " oi " << obc->obs.oi << dendl;

  OpContext *ctx = new OpContext(op, m->get_reqid(), &m->ops, obc, this);

  if (m->has_flag(CEPH_OSD_FLAG_SKIPRWLOCKS)) {
    dout(20) << __func__ << ": skipping rw locks" << dendl;
  } else if (m->get_flags() & CEPH_OSD_FLAG_FLUSH) {
    dout(20) << __func__ << ": part of flush, will ignore write lock" << dendl;
```

This gets assigned to OpContext::reqid which is then used to populate the rep_tid field on the repop:

```

  // issue replica writes
  ceph_tid_t rep_tid = osd->get_tid();

  RepGather *repop = new_repop(ctx, rep_tid);

  issue_repop(repop, ctx);
  eval_repop(repop);
```
PrimaryLogPG::issue_repop then passes repop->rep_tid as the tid:

```
 recovery_state.pre_submit_op(
    soid,
    ctx->log,
    ctx->at_version);
  pgbackend->submit_transaction(
    soid,
    ctx->delta_stats,
    ctx->at_version,
    std::move(ctx->op_t),
    recovery_state.get_pg_trim_to(),
    recovery_state.get_min_last_complete_ondisk(),
    std::move(ctx->log),
    ctx->updated_hset_history,
    on_all_commit,
    repop->rep_tid,
    ctx->reqid,
    ctx->op);
}
```

ReplicatedBackend::generate_subop then uses that tid to populate the MOSDRepOp:

```
Message * ReplicatedBackend::generate_subop(
  const hobject_t &soid,
  const eversion_t &at_version,
  ceph_tid_t tid,
  osd_reqid_t reqid,
  eversion_t pg_trim_to,
  eversion_t min_last_complete_ondisk,
  hobject_t new_temp_oid,
  hobject_t discard_temp_oid,
  const bufferlist &log_entries,
  std::optional<pg_hit_set_history_t> &hset_hist,
  ObjectStore::Transaction &op_t,
  pg_shard_t peer,
  const pg_info_t &pinfo)
{
  int acks_wanted = CEPH_OSD_FLAG_ACK | CEPH_OSD_FLAG_ONDISK;
  // forward the write/update/whatever
  MOSDRepOp *wr = new MOSDRepOp(
    reqid, parent->whoami_shard(),
    spg_t(get_info().pgid.pgid, peer.shard),
    soid, acks_wanted,
    get_osdmap_epoch(),
    parent->get_last_peering_reset_epoch(),
    tid, at_version);
```

PrimaryLogPG::record_write_error and PrimaryLogPG::submit_log_entries actually seem to use PrimaryLogPG::new_repop which uses osd->get_tid() rather than the tid on the client op.  I think this is actually a bug -- submit_log_entries predates error logging and the error logging pathway should really be using the client operation's tid.  @jdurgin Do you recall whether there's a good reason to be using a fresh tid for those operations?

Users of simple_opc_submit do need to generate their own OSD-global tid's since they are generated internally to the osd (watch expiration, snap trim, etc) rather than on behalf of a client operation.  The only similar user crimson has at the moment would be watch expiration, but I don't see the logic for that yet.  @rzarzynski Where's the watch expiration logic?

In summary, normal client IO and error log entries should use the tid from the original MOSDOp.

Reply:
It's a bug, it was not intended to use a new tid for error log entries.

Reply:
@athanatos there are two data structure, one is ceph_tid_t ( often named tid) another is osd_reqid_t (often named reqid which contains client tid), in classic osd, they use osd->get_tid() to create its own tid, and use OpContext->reqid (get from client request by m->get_reqid()) as reqid when create MOSDRepOp.  So it seems not a bug,  two id one stands for osd, one stands for client.  What do you think?

Reply:
MOSDRepOp *wr = new MOSDRepOp(
    reqid, parent->whoami_shard(),   // here reqid is get from OpContext->reqid get from m->get_reqid()
    spg_t(get_info().pgid.pgid, peer.shard),
    soid, acks_wanted,
    get_osdmap_epoch(),
    parent->get_last_peering_reset_epoch(),
    tid, at_version);    // here tid is ceph_tid_t get from osd->get_tid();

so seems reqid and tid shouldn't contain the same infomation.

Reply:
You are right!  I'm sorry, I managed to totally confuse myself.